src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py

```diff
@@ -226,17 +224,47 @@ def acr_update_set(client,
     return client.update(resource_group_name, registry_name, parameters)
 
 
-def acr_login(registry_name, resource_group_name=None, username=None, password=None):
+def _try_aad_token_login(login_server):
+    try:
+        username = "00000000-0000-0000-0000-000000000000"
+        password = get_login_refresh_token(login_server)
+        return username, password
+    except Exception as e:  # pylint: disable=broad-except
+        logger.warning("AAD authentication failed with message: %s", str(e))
+        return None, None
+
+
+def _try_admin_user_login(registry_name):
+    try:
+        cred = acr_credential_show(registry_name)
+        username = cred.username
+        password = cred.passwords[0].value
+        return username, password
+    except Exception as e:  # pylint: disable=broad-except
+        logger.warning("Admin user authentication failed with message: %s", str(e))
+        return None, None
+
+
+def acr_login(registry_name,
+              resource_group_name=None,
+              username=None,
+              password=None):
     """Login to a container registry through Docker.
     :param str registry_name: The name of container registry
     :param str resource_group_name: The name of resource group
     :param str username: The username used to log into the container registry
     :param str password: The password used to log into the container registry
     """
+    from subprocess import PIPE, Popen
+    docker_not_available = "Please verify whether docker is installed and running properly"
+
     try:
-        call(["docker", "ps"], stdout=PIPE, stderr=PIPE)
+        p = Popen(["docker", "ps"], stdout=PIPE, stderr=PIPE)
+        returncode = p.wait()
+        if returncode:
+            raise CLIError(docker_not_available)
```

Comment:
Exception is raised here and captured in the next `except` block and then throw again. Over-complicated flow.

Reply:
I suggest you __not__ to throw exception intentionally in try block, because if will capture in the following __bare exception__ and recapture and re-raise. This is wasteful.